### PR TITLE
[Autofix] ⚡ Feature: Host Google Fonts Locally

### LINKED ISSUE #436: ⚡ Feature: Host Google Fonts Locally
### Summary
Automatically detect Google Fonts loaded from fonts.googleapis.com, download them, and serve locally

### Why it matters
Eliminates external DNS lookups and HTTP requests to Google. Improves GDPR compliance (no data sent to Google). Reduces connection overhead. Offered by Perfmatters, FlyingPress, OMGF

### Current State
Plugin preloads fonts but doesn't intercept Google Fonts to serve locally

### Proposed Implementation
- Detect `fonts.googleapis.com` and `fonts.gstatic.com` references in enqueued styles
- Download font files (woff2) to `wp-content/cache/wppo/fonts/`
- Rewrite CSS to point to local font files
- Add `font-display: swap` automatically (already done for other fonts)
- Add toggle in File Optimization or Preload Settings tab
- Auto-refresh when font selection changes

### Acceptance Criteria
- Auto-detect Google Fonts, local hosting, CSS rewrite, font-display:swap, UI toggle, cache invalidation

### PR DESCRIPTION
**Pull request by nilesh32236**
### What Was Done

Added "Host Google Fonts Locally" feature that automatically detects Google Fonts loaded from `fonts.googleapis.com`, downloads the CSS and font files (woff2), caches them locally, and rewrites URLs — eliminating external DNS lookups, improving GDPR compliance, and applying `font-display: swap` automatically.

### Approach

Follows **Option A** (File Optimization tab + `style_loader_tag` interception): the feature is configured via a toggle in the CSS Optimization card, hooks into `style_loader_tag` at priority 9 (before minification at priority 10), and also performs buffer-level interception to catch `@import` and inline `<link>` patterns. Font files are cached under `wp-content/cache/wppo/fonts/` using hashed filenames. Cache is cleared on setting toggle and on general cache purge.

### Key Changes

- **`includes/class-google-fonts.php` (new)**: Core class with 4 public methods — `process_style_tag()` intercepts enqueued styles, `process_buffer()` catches `@import`/inline `<link>` patterns via regex, `download_and_rewrite()` fetches CSS+font files and rewrites URLs, `clear_font_cache()` deletes the font cache directory. Uses Chrome UA for woff2 responses. Converts v1 API URLs to v2 format. Calls existing `CSS::inject_font_display_swap()`.

- **`includes/class-main.php`**: Added `hostGoogleFontsLocally => false` to default settings (line 123), loaded the new class file (line 187), registered `style_loader_tag` filter at priority 9 (after minifyCSS block), added font cache invalidation on setting toggle in `on_settings_update()`.

- **`includes/class-cache.php`**: Added `use Google_Fonts` import, added buffer-level Google Fonts interception in `process_buffer_only()`, added `Google_Fonts::clear_font_cache()` call in `delete_all_cache_files()`.

- **`src/components/FileOptimization.js`**: Added `hostGoogleFontsLocally: false` default and a `SwitchField` toggle in the CSS Optimization card.

- **`src/components/__tests__/FileOptimization.test.js`**: Added 2 tests — renders switch unchecked, submits with `hostGoogleFontsLocally: true` when toggled.

## Files Changed

- `build/index.asset.php`
- `build/index.js`
- `build/tab-file-optimization.js`
- `includes/class-cache.php`
- `includes/class-main.php`
- `src/components/FileOptimization.js`
- `src/components/__tests__/FileOptimization.test.js`
- `vendor/composer/installed.php`

## Test Plan

- Automated tests were run and passed
- The fix was verified with project lint/typecheck commands

---

*🤖 Auto-generated by opencode-ai-reviewer from branch `autofix/issue-436`.*

### CODE REVIEW

File: includes/class-google-fonts.php [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+		}
+
+		/**
+		 * Normalize a Google Fonts URL — convert v1 API (css?) to v2 (css2?) format.
+		 *
+		 * @param string $url The raw URL.
+		 * @return string Normalized URL or empty string if not a Google Fonts URL.
+		 * @since 2.7.0
+		 */
+		private function normalize_google_fonts_url( $url ) {
```
Comment (nilesh32236) [inline includes/class-google-fonts.php:252]: **CRITICAL**: V1-to-v2 Google Fonts URL conversion is incomplete. The v1 API uses ':weight' syntax (e.g., 'family=Roboto:400,700') which differs from v2's '@' syntax ('family=Roboto:wght@400;700'). The code blindly passes v1 query params to the /css2 endpoint, which will fail for any URL specifying font weights or styles. This silently breaks font downloading for the most common Google Fonts use cases, and the user's pages still load fonts from Google's servers (GDPR non-compliance) despite the toggle being enabled.

> 💡 **How to fix:** Parse the 'family' parameter value in v1 URLs and convert weight/style syntax to v2 format. For simple cases (family name only, no weights), the current pass-through works. For weights/styles, convert ':400,700' to 'wght@400;700' and ':ital,wght@...' appropriately. Alternatively, avoid the conversion entirely and use the v1 API endpoint (/css) directly instead of converting to /css2.

```suggestion
private function normalize_google_fonts_url( $url ) {
    if ( false === strpos( $url, 'fonts.googleapis.com' ) ) {
        return '';
    }

    if ( false !== strpos( $url, '/css2' ) ) {
        return $url;
    }

    // For v1 URLs, keep using /css endpoint with same URL to avoid format conversion issues.
    return $url;
}
```

File: includes/class-google-fonts.php [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+
+		/**
+		 * Download a single font file from Google's CDN.
+		 *
+		 * @param string $url   The font file URL.
+		 * @param string $dest  Local destination path.
+		 * @return bool True on success, false on failure.
+		 * @since 2.7.0
+		 */
+		private function download_font_file( $url, $dest ) {
```
Comment (nilesh32236) [inline includes/class-google-fonts.php:276]: **IMPORTANT**: Race condition in download_font_file with streaming writes. Two concurrent requests for the same font file both pass the file_exists() check and start streaming downloads to the same destination path. If one download fails, it deletes the file that the other request may have successfully written, causing the font cache to be corrupted.

> 💡 **How to fix:** Use an atomic write pattern: download to a temporary file first, then rename to the final destination. Use a file lock or a marker file to signal that a download is in progress to prevent concurrent downloads of the same font.

```suggestion
private function download_font_file( $url, $dest ) {
    $tmp = $dest . '.tmp.' . wp_rand();
    $response = wp_remote_get(
        $url,
        array(
            'timeout'    => 30,
            'user-agent' => self::CHROME_UA,
            'stream'     => true,
            'filename'   => $tmp,
        )
    );

    if ( is_wp_error( $response ) || 200 !== wp_remote_retrieve_response_code( $response ) ) {
        if ( file_exists( $tmp ) ) {
            wp_delete_file( $tmp );
        }
        return false;
    }

    $result = rename( $tmp, $dest );
    if ( ! $result && file_exists( $tmp ) ) {
        wp_delete_file( $tmp );
    }

    return file_exists( $dest );
}
```

File: includes/class-cache.php [hunk 3]
```diff
@@ -958,6 +964,7 @@ private function delete_all_cache_files(): bool {
 			}
 
 			Used_CSS::delete_all_used_css();
+			Google_Fonts::clear_font_cache();
```
Comment (nilesh32236) [inline includes/class-cache.php]: **IMPORTANT**: Google_Fonts::clear_font_cache() is called on every full cache purge (delete_all_cache_files). This means every cache clear — triggered by permalink changes, theme switches, settings saves, plugin activate/deactivate — forces all Google Fonts to be re-downloaded on the next page load. This creates a burst of external HTTP requests (20-40+ for typical multi-weight font setups) and significantly slows the first uncached page view.

> 💡 **How to fix:** Make font cache clearing opt-in by removing it from delete_all_cache_files(). Instead, only clear font cache when the hostGoogleFontsLocally setting toggles (already handled in Main::on_settings_update) or add a separate 'Clear Font Cache' action button in the UI.

```suggestion
// Remove this line from delete_all_cache_files()
// Google_Fonts::clear_font_cache();
```

File: includes/class-cache.php [hunk 4]
```diff
@@ -400,6 +401,11 @@ private function process_buffer_only( $buffer ) {
 			$buffer = $image_optimisation->add_delay_load_img( $buffer );
 			$buffer = $image_optimisation->lazy_load_videos( $buffer );
 
+			// Host Google Fonts locally via buffer-level interception.
+			if ( ! empty( $this->options['file_optimisation']['hostGoogleFontsLocally'] ?? false ) ) {
```
Comment (nilesh32236) [inline includes/class-cache.php:424]: **MINOR**: A new Google_Fonts instance is created on every buffer process call, even though Main::setup_hooks already creates one for the style_loader_tag filter. This means two separate instances exist per page load, each with its own copy of options, file cache dir, etc.

> 💡 **How to fix:** Inject the existing Google_Fonts instance into the Cache class via a setter, similar to how Image_Optimisation is injected. Cache already has set_image_optimisation() as a precedent. Alternatively, store the Google_Fonts instance as a Main class property and pass it to Cache's constructor.

```suggestion
Inject the existing Google_Fonts instance into the Cache class via a setter, similar to how Image_Optimisation is injected. Cache already has set_image_optimisation() as a precedent. Alternatively, store the Google_Fonts instance as a Main class property and pass it to Cache's constructor.
```

File: src/components/__tests__/FileOptimization.test.js [hunk 5]
```diff
@@ -448,4 +448,54 @@ describe( 'FileOptimization Component', () => {
...
+
+		expect(
+			screen.getByLabelText( /Host Google Fonts Locally/i )
+		).toBeInTheDocument();
+		expect(
+			screen.getByLabelText( /Host Google Fonts Locally/i )
+		).not.toBeChecked();
+	} );
+
+	it( 'submits hostGoogleFontsLocally setting correctly', async () => {
```
Comment (nilesh32236) [inline src/components/__tests__/FileOptimization.test.js:468]: **MINOR**: The submit test for hostGoogleFontsLocally uses apiCall.mockResolvedValueOnce({ success: true }) without a 'message' field, unlike other tests in the same file that include it. While the component falls back to the default successMessage, the inconsistency means the test doesn't verify the notification message for the Google Fonts toggle path.

> 💡 **How to fix:** Include a 'message' field in the resolved value for consistency with other tests and to verify notification rendering.

**Review by nilesh32236 — COMMENTED** [review]
## Executive Summary

**Purpose:** Adds a 'Host Google Fonts Locally' feature that intercepts Google Fonts requests, downloads CSS and woff2 files, caches them locally, and rewrites URLs to eliminate external DNS lookups and improve GDPR compliance.

**Risk:** 🟡 MEDIUM — The implementation is generally sound and follows codebase conventions, but the v1-to-v2 Google Fonts URL conversion has a critical gap that silently breaks font downloading for users requesting specific font weights/styles, and there are race conditions in concurrent font file downloads.

---

## PR Review Summary

Well-structured feature with clean separation of concerns, graceful degradation, and dual interception (style_loader_tag + buffer). The main concerns are an incomplete v1-to-v2 URL conversion that silently fails for font variants, a streaming download race condition, and aggressive font cache clearing on every full cache purge. Tests cover the basic toggle/submit path but lack failure scenarios for the download step.

**Ready to merge?** false

**Reasoning:** The v1-to-v2 URL conversion silently fails for font weight/style variants — a core functionality gap that must be fixed before merge. Race condition in concurrent font downloads and aggressive font cache clearing on every full purge are also important concerns.

### Strengths

- **includes/class-google-fonts.php:88** — Clean separation of concerns — the Google_Fonts class is a standalone unit with clearly named public methods (process_style_tag, process_buffer, download_and_rewrite, clear_font_cache) that integrate naturally with existing plugin architecture.
- **includes/class-google-fonts.php:99** — Graceful degradation: when download_and_rewrite fails ('' === $local_url), the original tag is returned unchanged, ensuring fonts always load even when local hosting fails.
- **includes/class-cache.php:405** — Buffer-level interception acts as a safety net for @import and inline <link> patterns that bypass the style_loader_tag filter, ensuring comprehensive coverage.
- **includes/class-main.php:472** — Font cache invalidation on setting toggle correctly handles both enable and disable transitions, with proper state management.

**Comment by nilesh32236** [comment]
## 🤖 Autofix Review

**Status:** 🔍 Reviewing (iteration 2/3)

### Summary

The implementation follows the codebase's established patterns (singleton instance injected into Cache via setter, style_loader_tag interception at priority 9 before minification, buffer-level @import interception). All 5 previously reported issues have been fixed. One important edge case around empty downloaded font files and one minor unused constant remain. The JS component and tests are consistent with existing patterns.

### Issues Found
- **IMPORTANT:** `includes/class-google-fonts.php:297` — No validation that the downloaded font file is non-empty. If the remote server returns 200 with an empty or truncated body (transient error, early connection close), the empty temp file is renamed to the destination and cached permanently. file_exists() returns true for 0-byte files, so subsequent page loads reference an empty woff2 file, silently breaking @font-face rendering on the frontend until manual cache clear.
  > 💡 **How to fix:** Add filesize() validation after the HTTP response check. Delete the temp file and return false if the file is empty or missing.
<details><summary>Show suggested fix</summary>

```suggestion
if ( is_wp_error( $response ) || 200 !== wp_remote_retrieve_response_code( $response ) ) {
			if ( file_exists( $tmp ) ) {
				wp_delete_file( $tmp );
			}
			return false;
		}

		if ( ! file_exists( $tmp ) || 0 === filesize( $tmp ) ) {
			if ( file_exists( $tmp ) ) {
				wp_delete_file( $tmp );
			}
			return false;
		}

		$result = rename( $tmp, $dest );
```
</details>
- **MINOR:** `includes/class-google-fonts.php:50` — The private constant FONTS_API ('https://fonts.googleapis.com/css2') is defined but never referenced anywhere in the class. After the fix to keep v1 URLs on the /css endpoint, this constant became dead code. While harmless, it misleads maintainers into thinking the class uses the v2 API.
  > 💡 **How to fix:** Remove the unused FONTS_API constant.
- **MINOR:** `includes/class-cache.php:416` — The Google Fonts buffer-level interception (process_buffer) runs only inside process_buffer_only(), which executes only when the page cache is enabled. When enableCache is false but hostGoogleFontsLocally is on, @import in <style> tags and inline <link> patterns in templates are not intercepted. The style_loader_tag filter handles enqueued fonts (the common case), so this is a limited gap.
  > 💡 **How to fix:** Add Google Fonts buffer processing to the standalone output buffer paths (process_used_css_only / start_used_css_buffer) when hostGoogleFontsLocally is enabled and enableCache is disabled.

### Strengths
- ✅ **includes/class-google-fonts.php:284** — Atomic write pattern with temp-file-then-rename (wp_rand suffix, cleanup on failure) correctly resolves the concurrent download race condition for font files.
- ✅ **includes/class-main.php:307** — style_loader_tag filter at priority 9 correctly executes before minify_css at priority 10, ensuring Google Fonts URLs are rewritten to local paths before CSS minification.
- ✅ **includes/class-main.php:481** — Font cache invalidation is scoped to only the setting toggle (hostGoogleFontsLocally), not triggered on every full cache purge — preventing unnecessary re-downloads.
- ✅ **includes/class-google-fonts.php:260** — normalized_google_fonts_url() correctly keeps v1 API URLs on the /css endpoint instead of attempting risky v1-to-v2 format conversion, avoiding silently broken multi-weight font configurations.

### Iteration History
- 🔧 **Iteration 1:** Fix applied — 1 critical, 2 important
- ❌ **Iteration 2:** 0 critical, 1 important remaining

**Comment by nilesh32236** [comment]
## 🔧 Autofix Applied

**Iteration:** 2
**Files changed:** 3
**Commit:** `fix: autofix iteration 2`

### Changed Files
- `includes/class-google-fonts.php`
- `includes/class-main.php`
- `review-output.jsonl`

### Fix Details

### What Was Done

Fixed 3 remaining issues from the Host Google Fonts Locally PR review: added empty file validation after font download, removed unused dead code, and extended buffer-level Google Fonts interception to the standalone used-CSS buffer paths.

### Approach

Each fix was minimal and targeted:
1. **Empty file guard**: Added `filesize()` check after the streaming download completes — if the temp file doesn't exist or is 0 bytes, it's cleaned up and `false` is returned, preventing corrupted `@font-face` rendering from cached empty woff2 files.
2. **Dead code cleanup**: Removed the unused `FONTS_API` constant that became dead after v1 URLs were intentionally kept on the `/css` endpoint.
3. **Buffer coverage gap**: Added `google_fonts->process_buffer()` calls in both `process_used_css_only()` (WP 6.9+ path) and `process_used_css_capture()` (legacy path), so `@import`/inline `<link>` patterns are intercepted even when `enableCache` is disabled.

### Key Changes

- **`includes/class-google-fonts.php:303-309`** — Added `filesize()` validation after HTTP download in `download_font_file()` to reject empty/truncated files.
- **`includes/class-google-fonts.php:44-50`** — Removed unused `FONTS_API` private constant.
- **`includes/class-main.php:591-595`** — Added Google Fonts buffer processing to `process_used_css_only()` for WP 6.9+ standalone buffer path.
- **`includes/class-main.php:620-624`** — Added Google Fonts buffer processing to `process_used_css_capture()` for legacy standalone buffer path.


---

🤖 The fix agent has applied changes. The PR will be reviewed again on the next iteration.

File: includes/class-google-fonts.php [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+				$url,
+				array(
+					'timeout'    => 30,
+					'user-agent' => self::CHROME_UA,
+					'stream'     => true,
+					'filename'   => $tmp,
+				)
+			);
+
+			if ( is_wp_error( $response ) || 200 !== wp_remote_retrieve_response_code( $response ) ) {
```
Comment (nilesh32236) [inline includes/class-google-fonts.php:289]: **IMPORTANT**: No validation that the downloaded font file is non-empty. If the remote server returns 200 with an empty or truncated body (transient error, early connection close), the empty temp file is renamed to the destination and cached permanently. file_exists() returns true for 0-byte files, so subsequent page loads reference an empty woff2 file, silently breaking @font-face rendering on the frontend until manual cache clear.

> 💡 **How to fix:** Add filesize() validation after the HTTP response check. Delete the temp file and return false if the file is empty or missing.

```suggestion
if ( is_wp_error( $response ) || 200 !== wp_remote_retrieve_response_code( $response ) ) {
			if ( file_exists( $tmp ) ) {
				wp_delete_file( $tmp );
			}
			return false;
		}

		if ( ! file_exists( $tmp ) || 0 === filesize( $tmp ) ) {
			if ( file_exists( $tmp ) ) {
				wp_delete_file( $tmp );
			}
			return false;
		}

		$result = rename( $tmp, $dest );
```

**Review by nilesh32236 — COMMENTED** [review]
## Executive Summary

**Purpose:** Adds Host Google Fonts Locally feature that detects Google Fonts from fonts.googleapis.com, downloads CSS and woff2 files, caches them under wp-content/cache/wppo/fonts/, and rewrites URLs to local paths with font-display: swap applied.

**Risk:** 🟢 LOW — All 5 previously reported issues (critical v1→v2 URL bug, race condition, overly broad cache purge, duplicate instances, inconsistent test mock) have been correctly resolved. One important edge case (empty file validation for failed downloads) and one minor issue (unused constant) remain — both are straightforward fixes with no security or breaking concerns.

---

## PR Review Summary

The implementation follows the codebase's established patterns (singleton instance injected into Cache via setter, style_loader_tag interception at priority 9 before minification, buffer-level @import interception). All 5 previously reported issues have been fixed. One important edge case around empty downloaded font files and one minor unused constant remain. The JS component and tests are consistent with existing patterns.

**Ready to merge?** false

**Reasoning:** An important edge case persists: if the remote Google Fonts server returns 200 with an empty/truncated body, a 0-byte woff2 file is cached permanently via atomic rename, silently breaking @font-face rendering on the frontend until manual cache clear. The fix is a simple filesize() guard after download.

### Strengths

- **includes/class-google-fonts.php:284** — Atomic write pattern with temp-file-then-rename (wp_rand suffix, cleanup on failure) correctly resolves the concurrent download race condition for font files.
- **includes/class-main.php:307** — style_loader_tag filter at priority 9 correctly executes before minify_css at priority 10, ensuring Google Fonts URLs are rewritten to local paths before CSS minification.
- **includes/class-main.php:481** — Font cache invalidation is scoped to only the setting toggle (hostGoogleFontsLocally), not triggered on every full cache purge — preventing unnecessary re-downloads.
- **includes/class-google-fonts.php:260** — normalized_google_fonts_url() correctly keeps v1 API URLs on the /css endpoint instead of attempting risky v1-to-v2 format conversion, avoiding silently broken multi-weight font configurations.

### Issues

- **MINOR:** `includes/class-cache.php:416` — The Google Fonts buffer-level interception (process_buffer) runs only inside process_buffer_only(), which executes only when the page cache is enabled. When enableCache is false but hostGoogleFontsLocally is on, @import in <style> tags and inline <link> patterns in templates are not intercepted. The style_loader_tag filter handles enqueued fonts (the common case), so this is a limited gap.
  > 💡 **How to fix:** Add Google Fonts buffer processing to the standalone output buffer paths (process_used_css_only / start_used_css_buffer) when hostGoogleFontsLocally is enabled and enableCache is disabled.

File: includes/class-google-fonts.php [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+		 */
+		private const CHROME_UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36';
+
+		/**
+		 * Google Fonts CSS2 API endpoint.
+		 *
+		 * @var string
+		 * @since 2.7.0
+		 */
+		private const FONTS_API = 'https://fonts.googleapis.com/css2';
```
Comment (nilesh32236) [inline includes/class-google-fonts.php]: **MINOR**: The private constant FONTS_API ('https://fonts.googleapis.com/css2') is defined but never referenced anywhere in the class. After the fix to keep v1 URLs on the /css endpoint, this constant became dead code. While harmless, it misleads maintainers into thinking the class uses the v2 API.

> 💡 **How to fix:** Remove the unused FONTS_API constant.

**Review by nilesh32236 — COMMENTED** [review]
## Executive Summary

**Purpose:** Adds Host Google Fonts Locally feature that intercepts Google Fonts via style_loader_tag and buffer processing, downloads CSS + woff2 files to wp-content/cache/wppo/fonts/, rewrites URLs to local paths with font-display: swap, eliminating external DNS lookups and improving GDPR compliance.

**Risk:** 🟢 LOW — Feature is opt-in behind a toggle, gracefully degrades on failure (original URL preserved), reuses existing plugin patterns (WP_Http, file caching, buffer processing, dependency injection via setters), cache directory is within existing wppo structure, and all previously reported issues across three review iterations have been resolved.

---

## PR Review Summary

The implementation is thorough and production-ready. All 7 previously reported issues across three review iterations have been correctly resolved — including null-guard on preg_replace_callback, diagnostic logging on download failures, atomic font file writes, unchecked download return value (falls back to original URL on failure), scoped cache purge, singleton injection via setter, empty file validation, dead code removal, and buffer coverage in used-CSS-only paths. Integrates cleanly with existing codebase conventions (Image_Optimisation pattern, Log::add, Cache::setter). No new issues found.

**Ready to merge?** true

**Reasoning:** All previously reported critical, important, and minor issues have been fixed. The implementation is architecturally sound, follows existing codebase patterns, handles errors gracefully with fallback to original Google Fonts URLs, and has appropriate cache scoping. No remaining blockers.

### Strengths

- **includes/class-google-fonts.php:74** — Constructor correctly computes both filesystem path and URL from FONTS_CACHE_DIR in a single place, avoiding string duplication and keeping them in sync.
- **includes/class-google-fonts.php:91** — process_style_tag() is a clean filter callback with early returns for logged-in users, disabled setting, and non-Google URLs — no nested conditionals.
- **includes/class-google-fonts.php:104** — Google_Fonts is a typed class property instantiated once in Main's constructor and injected into Cache via set_google_fonts(), following the same established pattern as Image_Optimisation to avoid duplicate instances.
- **includes/class-google-fonts.php:123** — process_buffer() uses two targeted preg_replace_callback patterns for <link> and @import rules, each with proper error fallback returning the original match on failure.
- **includes/class-main.php:190** — set_google_fonts() setter is cleanly implemented on Cache. Buffer-level interception uses the injected instance with a null-safe fallback pattern consistent with Image_Optimisation.
- **includes/class-google-fonts.php:203** — preg_replace_callback in download_and_rewrite() correctly uses a backreference to match the same quote character in the original @font-face src URL, preventing mixed-quote mismatches.
- **includes/class-google-fonts.php:260** — normalize_google_fonts_url() correctly keeps v1 API URLs on the /css endpoint instead of attempting risky v1-to-v2 format conversion, avoiding silently broken font downloads for multi-weight configurations.
- **includes/class-google-fonts.php:284** — download_font_file() uses atomic temp-file-then-rename with wp_rand() suffix to prevent race conditions from concurrent requests, plus filesize validation to guard against empty/truncated downloads.
- **includes/class-main.php:307** — style_loader_tag filter is registered at priority 9 (before minification at 10 and criticalCSS defer at 10), correctly ordering the processing pipeline so Google Fonts URLs are localised before CSS minification.
- **includes/class-main.php:482** — Font cache clearing is scoped to fire only when the hostGoogleFontsLocally setting toggles (old_gf !== new_gf), avoiding unnecessary HTTP requests on general cache purge.
- **includes/class-main.php:588** — Buffer-level Google Fonts interception is correctly wired into both process_used_css_only() (WP 6.9+) and process_used_css_capture() (legacy), closing the coverage gap where @import patterns were missed when page cache is disabled.

**Comment by nilesh32236** [comment]
## ✅ Ready to Merge

All issues have been resolved in PR #460.

The review agent has approved this PR. A maintainer can merge it at their discretion.

### Summary

The implementation is thorough and production-ready. All 7 previously reported issues across three review iterations have been correctly resolved — including null-guard on preg_replace_callback, diagnostic logging on download failures, atomic font file writes, unchecked download return value (falls back to original URL on failure), scoped cache purge, singleton injection via setter, empty file validation, dead code removal, and buffer coverage in used-CSS-only paths. Integrates cleanly with existing codebase conventions (Image_Optimisation pattern, Log::add, Cache::setter). No new issues found.

**Comment by nilesh32236** [comment]
## ✅ Autofix PR Merged

PR #460 **[Autofix] ⚡ Feature: Host Google Fonts Locally** has been merged to `master`.

All review iterations completed and the reviewer agent confirmed the PR was ready.

---
*🤖 Merged automatically by the [AI Multi-Agent Review](https://github.com/nilesh32236/performance-optimisation/actions/runs/30460320598) workflow*
